File: python/cwe639_idor.py
from flask import Flask, jsonify

from models import Order  # SQLAlchemy model exposing a `.query` manager

app = Flask(__name__)


# CWE-639 — Authorization Bypass Through User-Controlled Key (IDOR).
# The user-controlled order id reaches a by-id lookup with no comparison to
# the authenticated principal (no ownership scoping).
@app.route("/orders/<order_id>")
def get_order(order_id):
    # SOURCE: user-controlled path parameter.
    # SINK (IDOR_LOOKUP): filter_by(id=...) with no ownership constraint.
    order = Order.query.filter_by(id=order_id).first()
    return jsonify(order.as_dict() if order else {})


if __name__ == "__main__":
    app.run(port=5001)

File: js/cwe639Idor.js
const express = require('express');

const app = express();

// naive in-memory data access — no ownership checks anywhere.
const invoices = {};
function getInvoiceById(id) {
  return invoices[id];
}

// CWE-639 — Authorization Bypass Through User-Controlled Key (IDOR).
// The user-controlled invoice id reaches a by-id lookup with no comparison
// to the authenticated session user.
app.get('/invoices/:invoiceId', (req, res) => {
  // SOURCE: user-controlled path parameter.
  const invoiceId = req.params.invoiceId;
  // SINK: generic by-id lookup, no ownership/authorization check.
  const invoice = getInvoiceById(invoiceId);
  res.json(invoice);
});

module.exports = app;
